File: python/6_ZigZag Conversion.py
class Solution {
public:
    string convert(string s, int numRows) {
        int st = 1, cur = 1, n = s.length(), next1 = 0, next2 = 0, i = 0;
        string res(s.length(),'_');
        if (numRows == 1) return s;
        ++i;
        res[i - 1] = s[st - 1];
        while (st <= numRows) {
            next1 = cur + 2 * numRows - 2 * st;
            next2 = next1 + 2 * (st - 1);
            if ((next1 != next2) && (cur != next1)) {
                if (next1 <= n) {
                    ++i;
                    res[i - 1] = s[next1 - 1];
                }
                else {
                    ++st;
                    if (st <= numRows) {
                        ++i;
                        res[i - 1] = s[st - 1];
                    }
                    cur = st;
                    continue;
                }
            }
            if (next2 <= n) {
                ++i;
                res[i - 1] = s[next2 - 1];
            }
            else {
                ++st;
                if (st <= numRows) {
                        ++i;
                        res[i - 1] = s[st - 1];
                }
                cur = st;
                continue;                
            }
            cur = next2;
        }
        return res;
    }
};

File: java/6_ZigZag Conversion.java
/**
 * 简单数学推导
 */
class Solution {
    public String convert(String s, int numRows) {
        int m = s.length();
        int cnt = 0;
        if (m == 0 || numRows <= 1) { return s; }
        StringBuilder resultSb = new StringBuilder();        
        for (int i = 0; i < numRows; ++i) {
            int col = 0;
            while (col - i < m) {
                int l = col - i;
                int r = col + i;
                if (l >= 0) { resultSb.append(s.charAt(l)); }
                if (r != l && r < m && i != numRows - 1) { resultSb.append(s.charAt(r)); }
                col += (2 * numRows - 2);
            }
        }
        return resultSb.toString();
    }
}
/**Start
    0       2k-2        // n * (2k - 2)
    1      2k               n * (2k-2)  -1/+1
    2    ..
    3   k+1
    .. k
    k-1
   **/     
    